app/api/csv/csvExporter.ts

```diff
@@ -0,0 +1,217 @@
+import { Writable } from 'stream';
+import { EventEmitter } from 'events';
+import * as csv from '@fast-csv/format';
+import templates from 'api/templates';
+import { PropertySchema } from 'shared/types/commonTypes';
+import { TemplateSchema } from 'shared/types/templateType';
+import translate, { getLocaleTranslation, getContext } from 'shared/translate';
+import translations from 'api/i18n/translations';
+import formatters from './typeFormatters';
+
+export type SearchResults = {
+  rows: any[];
+  totalRows: number;
+  aggregations: {
+    all: {
+      _types: {
+        buckets: {
+          key: string;
+          // eslint-disable-next-line camelcase
+          doc_count: number;
+          // eslint-disable-next-line camelcase
+          filtered: { doc_count: number };
+        }[];
+      };
+    };
+  };
+};
+
+export type ExporterOptions = {
+  dateFormat: string;
+  language: string;
+};
+
+export type ExportHeader = {
+  label: string;
+  name: string;
+  common: boolean;
+};
+
+export type TemplatesCache = {
+  [id: string]: TemplateSchema;
+};
+
+export const getTypes = (searchResults: SearchResults, typesWhitelist: string[] = []) =>
+  typesWhitelist.length > 0
+    ? typesWhitelist
+    : searchResults.aggregations.all._types.buckets
+        .filter(bucket => bucket.filtered.doc_count > 0)
+        .map(bucket => bucket.key);
+
+const hasValue = (value: string) => value !== 'missing';
+
+export const getTemplatesModels = async (templateIds: string[]): Promise<TemplatesCache> =>
+  Promise.all(
+    templateIds.filter(hasValue).map(async (id: string) => templates.getById(id))
+  ).then(results =>
+    results.reduce<any>(
+      (memo, template) => (template ? { ...memo, [template._id]: template } : memo),
+      {}
+    )
+  );
+
+export const notDuplicated = (collection: any) => (item: any) =>
+  collection.findIndex((i: any) => Object.keys(i).every(key => i[key] === item[key])) < 0;
+
+export const excludedProperties = (property: PropertySchema) =>
+  !['geolocation', 'preview', 'markdown', 'nested'].includes(property.type);
+
+export const processHeaders = (templatesCache: TemplatesCache): ExportHeader[] =>
+  Object.values(templatesCache).reduce(
+    (memo: any[], template: any) =>
+      memo.concat(
+        template.properties
+          .filter(excludedProperties)
+          .map((property: any) => ({ label: property.label, name: property.name }))
+          .filter(notDuplicated(memo))
+      ),
+    []
+  );
+
+export const prependCommonHeaders = (headers: ExportHeader[]) =>
+  [
+    {
+      label: 'Title',
+      name: 'title',
+      common: true,
+    },
+    {
+      label: 'Date added',
+      name: 'creationDate',
+      common: true,
+    },
+    {
+      label: 'Template',
+      name: 'template',
+      common: true,
+    },
+  ].concat(headers);
+
+export const concatCommonHeaders = (headers: ExportHeader[]) =>
+  headers.concat([
+    { label: 'Geolocation', name: 'geolocation', common: true },
```

Comment:
While Documents, Attachments and Published are 'higher order' properties, the `_geolocation` properties reside within the metadata, they are not common to all templates, and there can be several with different labels for a single template.  I'm having a hard time following the process of geolocation.  I think there is something wrong in the concept here.

Reply:
The geolocation property is treated as a special case because I found there is some different logic going on in its displaying across the app. I may be totally wrong and had misunderstood the concept. Let's have a chat asap.